tests/test_noise.py

```diff
@@ -0,0 +1,142 @@
+from __future__ import annotations
+
+from typing import Callable
+
+import jax.numpy as jnp
+import numpy as np
+import pytest
+
+from horqrux.api import expectation, run, sample
+from horqrux.apply import apply_gate
+from horqrux.noise import NoiseInstance, NoiseType
+from horqrux.parametric import PHASE, RX, RY, RZ
+from horqrux.primitive import NOT, H, I, S, T, X, Y, Z
+from horqrux.utils import ForwardMode, density_mat, product_state, random_state
+
+MAX_QUBITS = 7
+PARAMETRIC_GATES = (RX, RY, RZ, PHASE)
+PRIMITIVE_GATES = (NOT, H, X, Y, Z, I, S, T)
+
+NOISE_oneproba = (
+    NoiseType.BITFLIP,
+    NoiseType.PHASEFLIP,
+    NoiseType.DEPOLARIZING,
+    NoiseType.AMPLITUDE_DAMPING,
+    NoiseType.PHASE_DAMPING,
+)
+ALL_NOISES = list(NoiseType)
+
+
+def noise_instance(noise_type: NoiseType) -> NoiseInstance:
+    if noise_type in NOISE_oneproba:
+        errors = 0.1
+    elif noise_type == NoiseType.PAULI_CHANNEL:
+        errors = (0.4, 0.5, 0.1)
+    else:
+        errors = (0.2, 0.8)
+
+    return NoiseInstance(noise_type, error_probability=errors)
+
+
+@pytest.mark.parametrize("gate_fn", PRIMITIVE_GATES)
+@pytest.mark.parametrize("noise_type", ALL_NOISES)
+def test_noisy_primitive(gate_fn: Callable, noise_type: NoiseType) -> None:
```

Comment:
I would also add test to check if the safeguards you've added are raised succesfully.